### ISSUE TITLE
SocketException: Failed host lookup

### ISSUE THREAD

**Issue by yegor-pelykh** [flutter/flutter#132912]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Just make a network request.

For example, using the basic `http` package.

Also relevant when using other packages, such as `Dio`.
The same error, and in the same versions of Flutter.

Therefore, I think they have the same root.

**The bug is very critical, as already many communities of various applications complain about it.**

I have given the simplest code example below.
The main essence is in the line:
```dart
await http.get(Uri.parse('https://flutter.dev'));
```

This request throws an exception:
`SocketException (SocketException: Failed host lookup: 'flutter.dev' (OS Error: ... , errno = 11004))`

This error appears on ANY host of the many that I have tried.
That is, the problem is not in a particular host.

This problem is relevant for **Windows** (which discards many other similar problems that can be found on the Internet and which are associated with missing permissions on the Android system). It's not that problem.

I've reproduced this bug in the **latest master** as well as in the **latest stable release**.
On the Internet people advise to downgrade Flutter to version `3.10.5`, where this problem does not reproduce. I personally haven't tried it yet, but I suspect that all versions after `3.10.5` have this bug.
But downgrading to that version is not an option for me anyway.

Please fix this bug as soon as possible.
Yes, this exception is skippable, but it creates debugging inconvenience in many cases.
Moreover, this problem is relevant for many.

Thank you!

### Expected results

No exception

### Actual results

This request throws an exception:
`SocketException (SocketException: Failed host lookup: 'flutter.dev' (OS Error: ... , errno = 11004))`

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Future<void> _test() async {
    try {
      await http.get(Uri.parse('https://flutter.dev'));
    } catch (ex) {
      // ignore: use_build_context_synchronously
      showDialog(
        context: context,
        builder: (context) {
          return AlertDialog(
            title: const Text('Result'),
            content: Text(
              ex.toString(),
            ),
            actions: <Widget>[
              TextButton(
                child: const Text('Ok'),
                onPressed: () {
                  Navigator.of(context).pop();
                },
              ),
            ],
          );
        },
      );
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      floatingActionButton: FloatingActionButton(
        onPressed: _test,
        child: const Icon(Icons.error),
      ),
    );
  }
}

```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.14.0-8.0.pre.17, on Microsoft Windows [Version 10.0.22621.2134], locale ru-UA)
    • Flutter version 3.14.0-8.0.pre.17 on channel master at C:\SDK\Flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6f227c0784 (2 days ago), 2023-08-18 19:04:15 -0700
    • Engine revision f4bffdcf85
    • Dart version 3.2.0 (build 3.2.0-87.0.dev)
    • DevTools version 2.26.1

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at C:\SDK\Android
    • Platform android-33-ext5, build-tools 33.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files\Chromium\chrome

[✓] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.7.1)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.7.34009.444
    • Windows 10 SDK version 10.0.22621.0

[✓] Android Studio (version 2022.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-9586694)

[✓] VS Code (version 1.81.1)
    • VS Code at C:\Users\Yegor\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.70.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22621.2134]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 115.0.1901.203

[✓] Network resources
    • All expected network resources are available.
```

</details>

**Comment by arsabe75** [comment]
I have a similar issue with Image.network(). This issue happens only when debugging (note: can run OK once, but if I hot restart the aplication the issue appears). If I run the application without debugging all works OK or if I use VCode and dissable Uncaught exceptions I can run the application without this issue.
![error failed host lookup](https://github.com/flutter/flutter/assets/54506291/6c1ae53f-f6f0-4729-a9aa-59d574d755a1)



![working OK](https://github.com/flutter/flutter/assets/54506291/ad98fcfb-7a87-41bd-8399-d6321c04e4b7)

**Comment by zhxst** [comment]
Similar issue here `SocketException (SocketException: Failed host lookup: 'xxxxx' (OS Error: No address associated with hostname, errno = 7))`.
It was thrown at _NativeSocket.lookup from socket_patch.dart, and the previous one was _NativeSocket.staggeredLookup, but didn't have a way to track it down. because of the async gap.
The exceptions were thrown three times in a row, and repeated every time a network request was sent.

## Flutter Doctor output
[✓] Flutter (Channel stable, 3.13.0, on macOS 13.4.1 22F770820d darwin-arm64, locale zh-Hans-CN)
    • Flutter version 3.13.0 on channel stable at /Users/zhx/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (5 days ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/zhx/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode_14.3.1.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.71.20230801

[✓] Connected device (2 available)
    • MI 8 (mobile)   • ba9b410d • android-arm64 • Android 10 (API 29)
    • macOS (desktop) • macos    • darwin-arm64  • macOS 13.4.1 22F770820d darwin-arm64

[✓] Network resources
    • All expected network resources are available.

**Comment by yegor-pelykh** [comment]
I also opened an issue like this in the `http` package repository:
#https://github.com/dart-lang/http/issues/1008

**Comment by darshankawar** [comment]
@yegor-pelykh 
Can you take a look at [this](https://github.com/flutter/flutter/issues/129121) issue with underlying comments and check if it resembles your case or not ?

Also check https://github.com/Dart-Code/Dart-Code/issues/4695#issuecomment-1685941945 and see if it helps further for reference or not.

**Comment by yegor-pelykh** [comment]
Hi @darshankawar 
I looked at the comment, but I didn't see how it could help with this problem.
If you mean disabling something like turning off "all exceptions" or "unhandled exceptions" as a solution, then I know about it, and so far I'm doing it locally.
But this is just a temporary solution and it is not suitable as a permanent fix. It should not be necessary to turn off all unhandled exceptions so that the part of the code responsible for the DNS lookup in flutter does not interfere with development.
Therefore, I see the correct fix only to get rid of this exception without changing the IDE settings.

**Comment by darshankawar** [comment]
Thanks for the update @yegor-pelykh 
I am unable to replicate this using the code sample you provided and running with AS. As you mentioned it is relevant on Windows, can you try [this](https://github.com/flutter/flutter/issues/129121#issuecomment-1609321825) and check if you are consistently able to replicate the SocketException ? That code sample is without Flutter code and just uses dart core code.

**Comment by koodimetsa** [comment]
This seems to be happening only for some domains. In example below test 1 works, but test 2 end up in error:
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

```
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: SafeArea(
            child: Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        TextButton(onPressed: _testPress1, child: const Text('Test press 1')),
        TextButton(onPressed: _testPress2, child: const Text('Test press 2')),
      ],
    )));
  }

  _testPress1() async {
    final url = Uri.parse('https://yle.fi');

    final response = await http.get(
      url,
    );

    print(response.body);
  }

  _testPress2() async {
    final url = Uri.parse('https://www.mtvuutiset.fi');

    final response = await http.get(
      url,
    );

    print(response.body);
  }
}

```

**Comment by arhamcode** [comment]
i found this exception raised when we made request to website that doesn't ready for IPv6. For example in sample code by @koodimetsa :
- https://yle.fi (is ready for IPv6)
- https://www.mtvuutiset.fi (not ready for IPv6) will throw SocketException

Related closed issue here [#50868](https://github.com/dart-lang/sdk/issues/50868) and has been fixed on iOS, maybe, thats why it only appear on android.

**Comment by jmpavan** [comment]
In my case the problem was given by the last stable version 3.13.0: I did downgrade to 3.10.6 and the problem disappeared.

**Comment by azizkazdal** [comment]
I have same issue. Downgrade to 3.10.6. Need to fix.

**Comment by darshankawar** [comment]
Revisiting this issue based on several reports, we verified this on Windows + Android setup using vscode and were able to replicate the error in following case:

- In Run without debug mode in VSCode: No issue
- In Debug mode, it's reproducible with below error
`OSError (OS Error: No address associated with hostname, errno = 7)`

Thanks @huycozy for verifying.

But since, `SocketException` comes from `dart: io` library, we are inclined towards saying that the issue probably belongs to dart-lang repo and also based on [this](https://github.com/flutter/flutter/issues/132912#issuecomment-1688593158) verification.

So @yegor-pelykh and everyone facing this issue, I suggest to file this issue in https://github.com/dart-lang/sdk/issues.
I'll close from here. If anybody disagrees, write in comments and I'll reopen it for further action.

**Comment by KostyaLocal** [comment]
@arhamcode 
This error occurs not only if there is no ipv6 on the server, but also if there is no ipv6 support on the client side

So, whats the current solution? back to 3.10.6?

**Comment by jmpavan** [comment]
At least in my case in 3.10.6 the problem does not occur.

Jorge Mario Paván
+54 9 11 5327 9247
***@***.***

**Comment by KostyaLocal** [comment]
@jmpavan yes, this was broken in 3.13. 

Found another temporary solution - configure ipV6 on the target server and write a AAAA-record for target host in the domain zone. However, this won't always work - for example if there is no access to the domain zone (not your domain), or ipV6 doesn't work on the local network.

**Comment by moshOntong-IT** [comment]
I got also an error in image network in flutter windows desktop

```
SocketException (SocketException: Failed host lookup: 'cloud.appwrite.io' (OS Error: No such host is known.
, errno = 11001))
```

But when I continue the debug, it success loaded the image. I dont know why this SocketException occurs

**Comment by brianquinlan** [comment]
This really looks like https://github.com/flutter/flutter/issues/129121

**Comment by zhxst** [comment]
It's fixed on Flutter 3.13.4 / Dart 3.1.2
Finally I can upgrade my app to 3.13 !

**Comment by scottparise** [comment]
it's not fixed on Flutter 3.13.5 - still have the same issue

**Comment by busslina** [comment]
I had two uncaught exception, then the third I'm able to catch it. Windows
